pkg/abstractions/common/ring_buffer.go

```diff
@@ -20,19 +20,34 @@ func NewRingBuffer[T any](size int) *RingBuffer[T] {
 	}
 }
 
-// Push adds a new request to the buffer
-func (rb *RingBuffer[T]) Push(request T) {
+// Push adds a new request to the buffer. If priority is true, it gets inserted at the front of the buffer
+func (rb *RingBuffer[T]) Push(request T, priority bool) {
 	rb.mu.Lock()
 	defer rb.mu.Unlock()
 
-	if rb.count == rb.size {
-		rb.head = (rb.head + 1) % rb.size
+	if priority {
+		if rb.count == rb.size {
+			// Buffer is full, overwrite the oldest element at head without moving pointers
+			// No adjustment to head or tail, since the count remains the same
+			rb.buffer[rb.head] = request
+		} else {
+			// Move head backward to insert at front
+			rb.head = (rb.head - 1 + rb.size) % rb.size
+			rb.buffer[rb.head] = request
+			rb.count++
+		}
 	} else {
-		rb.count++
-	}
+		// Normal FIFO insert at tail
+		rb.buffer[rb.tail] = request
+		rb.tail = (rb.tail + 1) % rb.size
 
-	rb.buffer[rb.tail] = request
-	rb.tail = (rb.tail + 1) % rb.size
+		// Buffer is full, move head forward to overwrite oldest element
+		if rb.count == rb.size {
+			rb.head = (rb.head + 1) % rb.size
```

Comment:
Also if we do overwrite it, it doesn't seem to cancel the users request. The request needs to be canceled or the user client will be left hanging

Reply:
Yeah but it's the same as the fifo implementation was previously. Since it's a ring buffer it would always overwrite the first element when it's full. In practice it won't ever be full because the buffer size is likely much larger than max pending tasks